### PR TITLE
gnrc_ipv6: make gnrc_netapi documentation more detailed

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Now that #10233 is merged the documentation can be made in a more detailed fashion, since a lot of special cases do not exist anymore (and thus would just explode the specification). I followed the style laid out in `net/gnrc/sixlowpan.h` (see #10582 for [unrelated] updates to that).

### Testing procedure
Read the doc, see if you can understand it.

### Issues/PRs references
None (though a somewhat follow-up to #10233).

### CODE REVIEW

File: sys/include/net/gnrc/ipv6.h [hunk 1]
```diff
@@ -11,10 +11,63 @@
...
+ *    extension header between the IPv6 header and the payload marked as
+ *    separate @ref gnrc_pktsnip_t (i.e. the IPv6 payload will be the first
+ *    snip due to receive order). Note, that IPv6-internal headers (such as
+ *    ICMPv6 and extension headers) are handled within this module.
+ * 1. If the network interface the packet was received over (indicated by the
+ *    gnrc_netif_hdr_t::if_pid if provided or the destination address in the
+ *    IPv6 header) is configured as a forwarding interface, the destination is
+ *    not an address on this host, and a route is configured for the destination
+ *    the packet will be forwarded. This means the module will emit a
+ *    `GNRC_NETAPI_MSG_TYPE_SND` with the packet reversed into send order.
```
Comment (smlng) [inline sys/include/net/gnrc/ipv6.h]: does that mean each packet is recreated before being forwarded? That sound rather inefficient ...

File: sys/include/net/gnrc/ipv6.h [hunk 2]
```diff
@@ -11,10 +11,63 @@
  * @ingroup     net_gnrc
  * @brief       GNRC's IPv6 implementation
  *
- * The IPv6 control thread understands messages of type
- *
- *  * @ref GNRC_NETAPI_MSG_TYPE_RCV, and
- *  * @ref GNRC_NETAPI_MSG_TYPE_SND,
+ * This module is for usage with the @ref net_gnrc_netapi
+ *
+ * # NETAPI command documentation
```
Comment (smlng) [inline sys/include/net/gnrc/ipv6.h]: shouldn't this be `IPv6 NETAPI command documentation`?

File: sys/include/net/gnrc/ipv6.h [hunk 3]
```diff
@@ -11,10 +11,63 @@
...
+ *    application is registered to the next header field of the IPv6 header,
+ *    the next header field of one of the IPv6 extension headers within the
+ *    packet or @ref gnrc_nettype_from_protnum() resolves one of the previously
+ *    mentioned next header fields to a @ref net_gnrc_nettype, the packet will
+ *    be forwarded to that application, with the IPv6 header and every IPv6
+ *    extension header between the IPv6 header and the payload marked as
+ *    separate @ref gnrc_pktsnip_t (i.e. the IPv6 payload will be the first
+ *    snip due to receive order). Note, that IPv6-internal headers (such as
+ *    ICMPv6 and extension headers) are handled within this module.
+ * 1. If the network interface the packet was received over (indicated by the
```
Comment (smlng) [inline sys/include/net/gnrc/ipv6.h]: `1.` after `2.` doesn't matter when rendered, but as plain text ....

File: sys/include/net/gnrc/ipv6.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: No, if the space is available the snips are just reverted. Also unrelated to this PR, since this is a doc update.

File: sys/include/net/gnrc/ipv6.h (hunk 2 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Why should it?

File: sys/include/net/gnrc/ipv6.h (hunk 3 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Oops.

File: sys/include/net/gnrc/ipv6.h (hunk 2 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Ping @smlng?

File: sys/include/net/gnrc/ipv6.h [hunk 4]
```diff
@@ -11,10 +11,63 @@
...
- *  * @ref GNRC_NETAPI_MSG_TYPE_SND,
+ * This module is for usage with the @ref net_gnrc_netapi
+ *
+ * # NETAPI command documentation
+ *
+ * ## `GNRC_NETAPI_MSG_TYPE_RCV`
+ *
+ * @ref GNRC_NETAPI_MSG_TYPE_RCV expects a @ref net_gnrc_pkt (referred to as
+ * "packet" in the following) in receive order (payload first, headers in
+ * decreasing order by layer). It should at least contain a link-layer payload.
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: Does this mean: "The link-layer must contain a non-empty payload"?

File: sys/include/net/gnrc/ipv6.h [hunk 5]
```diff
@@ -11,10 +11,63 @@
...
- *  * @ref GNRC_NETAPI_MSG_TYPE_RCV, and
- *  * @ref GNRC_NETAPI_MSG_TYPE_SND,
+ * This module is for usage with the @ref net_gnrc_netapi
+ *
+ * # NETAPI command documentation
+ *
+ * ## `GNRC_NETAPI_MSG_TYPE_RCV`
+ *
+ * @ref GNRC_NETAPI_MSG_TYPE_RCV expects a @ref net_gnrc_pkt (referred to as
+ * "packet" in the following) in receive order (payload first, headers in
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: maybe better "headers ordered down the stack"

File: sys/include/net/gnrc/ipv6.h [hunk 6]
```diff
@@ -11,10 +11,63 @@
...
+ * decreasing order by layer). It should at least contain a link-layer payload.
+ * If it contains a @ref net_gnrc_netif_hdr it will be taken into account.
+ *
+ * If the link-layer payload is not an IPv6 packet (starting with a 0x6
+ * half-byte) or the module is not able to parse it correctly it will be
+ * dropped. If the packet can't be forwarded (either to another node or an
+ * application) it will be dropped. Otherwise, there are two scenarios of
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: Unclear: netapi is not forwarding to nodes, but stack-internal. Do you mean: "forwarded up or down the stack"?

File: sys/include/net/gnrc/ipv6.h [hunk 7]
```diff
@@ -11,10 +11,63 @@
...
+ * success:
+ *
+ * 1. If the destination address is an address on this host, and a suitable
+ *    application is registered to the next header field of the IPv6 header,
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: What is a "suitable application"? Who checks its appropriateness?

File: sys/include/net/gnrc/ipv6.h [hunk 8]
```diff
@@ -11,10 +11,63 @@
...
+ *    the next header field of one of the IPv6 extension headers within the
+ *    packet or @ref gnrc_nettype_from_protnum() resolves one of the previously
+ *    mentioned next header fields to a @ref net_gnrc_nettype, the packet will
+ *    be forwarded to that application, with the IPv6 header and every IPv6
+ *    extension header between the IPv6 header and the payload marked as
+ *    separate @ref gnrc_pktsnip_t (i.e. the IPv6 payload will be the first
+ *    snip due to receive order). Note, that IPv6-internal headers (such as
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: This long sentence is hard to parse. Can this be simplified to become one step one sentence?

File: sys/include/net/gnrc/ipv6.h [hunk 9]
```diff
@@ -11,10 +11,63 @@
...
+ *    ICMPv6 and extension headers) are handled within this module.
+ * 2. If the network interface the packet was received over (indicated by the
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: maybe better: "If the receiving network interface ..."

File: sys/include/net/gnrc/ipv6.h [hunk 10]
```diff
@@ -11,10 +11,63 @@
...
+ *    gnrc_netif_hdr_t::if_pid if provided or the destination address in the
+ *    IPv6 header) is configured as a forwarding interface, the destination is
+ *    not an address on this host, and a route is configured for the destination
+ *    the packet will be forwarded. This means the module will emit a
+ *    `GNRC_NETAPI_MSG_TYPE_SND` with the packet reversed into send order.
+ *
+ * ## `GNRC_NETAPI_MSG_TYPE_SND`
+ *
+ * @ref GNRC_NETAPI_MSG_TYPE_SND expects a @ref net_gnrc_pkt(referred to as
+ * "packet" in the following) in send order (headers first, payload last). It
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: headers ordered up the stack?

File: sys/include/net/gnrc/ipv6.h [hunk 11]
```diff
@@ -11,10 +11,63 @@
...
+ * should at least contain a snip of type @ref GNRC_NETTYPE_IPV6 as its first or
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: should?

File: sys/include/net/gnrc/ipv6.h [hunk 12]
```diff
@@ -11,10 +11,63 @@
...
+ * second snip. For the latter case the first snip must be of type @ref
+ * GNRC_NETTYPE_NETIF.
+ *
+ * If the destination address within the @ref GNRC_NETTYPE_IPV6 snip is an
+ * address on this host, the packet will be reversed and merged as though it
+ * came from a network interface (see previous section).
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: shouldn't it be "as if it came from the destination network interface" meaning, the packet undergoes an IP address re-writing? If not, the packet will carry the wrong address which will break the socket endpoint.
Side remark: Unclear, whether this is a comment to the doc or the actual code (I didn't check).

File: sys/include/net/gnrc/ipv6.h [hunk 13]
```diff
@@ -11,10 +11,63 @@
...
+ *
+ * Otherwise, if a route (or neighbor cache entry) to the destination address
+ * exists, the IPv6 header will be filled for fields that were not set by upper
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: "for fields" -> "with fields". Mention, where the values of these fields come from?

File: sys/include/net/gnrc/ipv6.h [hunk 14]
```diff
@@ -11,10 +11,63 @@
...
+ * layers and forwarded to that node.
```
Comment (tcschmidt) [inline sys/include/net/gnrc/ipv6.h]: Mhmm, "to that node"? Rather transferred to the appropriate L2 IF?

File: sys/include/net/gnrc/ipv6.h [hunk 15]
```diff
@@ -11,10 +11,63 @@
...
  * @brief       GNRC's IPv6 implementation
  *
- * The IPv6 control thread understands messages of type
- *
- *  * @ref GNRC_NETAPI_MSG_TYPE_RCV, and
- *  * @ref GNRC_NETAPI_MSG_TYPE_SND,
+ * This module is for usage with the @ref net_gnrc_netapi
+ *
+ * # NETAPI command documentation
+ *
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6.h]: A short introductory sentence a la "This module handles the @ref net_gnrc_netapi commands as follows:" would be nice :)

File: sys/include/net/gnrc/ipv6.h [hunk 16]
```diff
@@ -11,10 +11,63 @@
...
+ * ## `GNRC_NETAPI_MSG_TYPE_RCV`
+ *
+ * @ref GNRC_NETAPI_MSG_TYPE_RCV expects a @ref net_gnrc_pkt (referred to as
+ * "packet" in the following) in receive order (payload first, headers in
+ * decreasing order by layer). It should at least contain a link-layer payload.
+ * If it contains a @ref net_gnrc_netif_hdr it will be taken into account.
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6.h]: Writing "if it contains a @ref net_gnrc_netif_hdr, this header will [...]" could prevent any confusion whether this means that the header is taken into account if present or that the packet is taken into account *only* if this header is present

File: sys/include/net/gnrc/ipv6.h (hunk 6 above)
Comment (Lotterleben) [reply sys/include/net/gnrc/ipv6.h]: I'm assuming that "If the packet can't be forwarded" refers to the white/blacklisting checks in `gnrc.ipv6.c:_receive()`– if a packet's source address is blacklisted/not whitelisted, the packet **cannot** be forwarded. It doesn't matter if the ipv6 module is the module doing the actual forwarding; the ipv6 module is where this decision is made.
In case this assumption is correct, I'm happy with the sentence as it is– it states what happens without going into unnecessary detail.

File: sys/include/net/gnrc/ipv6.h [hunk 17]
```diff
@@ -11,10 +11,63 @@
...
+ *    be forwarded to that application, with the IPv6 header and every IPv6
+ *    extension header between the IPv6 header and the payload marked as
+ *    separate @ref gnrc_pktsnip_t (i.e. the IPv6 payload will be the first
+ *    snip due to receive order). Note, that IPv6-internal headers (such as
+ *    ICMPv6 and extension headers) are handled within this module.
+ * 2. If the network interface the packet was received over (indicated by the
+ *    gnrc_netif_hdr_t::if_pid if provided or the destination address in the
+ *    IPv6 header) is configured as a forwarding interface, the destination is
+ *    not an address on this host, and a route is configured for the destination
+ *    the packet will be forwarded. This means the module will emit a
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6.h]:
```suggestion
 *    the packet will be forwarded to. This means the module will emit a
```

Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6.h]:
```suggestion
 *    the packet will be forwarded. In consequence, the module will emit a
```

File: sys/include/net/gnrc/ipv6.h [hunk 18]
```diff
@@ -11,10 +11,63 @@
...
+ *    `GNRC_NETAPI_MSG_TYPE_SND` with the packet reversed into send order.
+ *
+ * ## `GNRC_NETAPI_MSG_TYPE_SND`
+ *
+ * @ref GNRC_NETAPI_MSG_TYPE_SND expects a @ref net_gnrc_pkt(referred to as
+ * "packet" in the following) in send order (headers first, payload last). It
+ * should at least contain a snip of type @ref GNRC_NETTYPE_IPV6 as its first or
+ * second snip. For the latter case the first snip must be of type @ref
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6.h]:
```suggestion
 * second snip. If the the first snip is not of type @ref GNRC_NETTYPE_IPV6, it must be of type @ref
```

File: sys/include/net/gnrc/ipv6.h [hunk 19]
```diff
@@ -11,10 +11,63 @@
...
+ * GNRC_NETTYPE_NETIF.
+ *
+ * If the destination address within the @ref GNRC_NETTYPE_IPV6 snip is an
+ * address on this host, the packet will be reversed and merged as though it
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6.h]: merged?

File: sys/include/net/gnrc/ipv6.h (hunk 4 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: No. Will try to clarify.

File: sys/include/net/gnrc/ipv6.h (hunk 5 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Mh... I think that phrasing is more confusing. Receive order at least is defined in http://doc.riot-os.org/group__net__gnrc__pkt.html#ga278e783e56a5ee6f1bd7b81077ed82a7.

File: sys/include/net/gnrc/ipv6.h (hunk 4 above)
Comment (tcschmidt) [reply sys/include/net/gnrc/ipv6.h]: Then maybe put the stronger phrase: "The link-layer must contain at least a valid IPv6 header".

File: sys/include/net/gnrc/ipv6.h (hunk 6 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: This does not refer to netapi but the packet level. If the packet can't be forwarded (i.e. there is no FIB entry) then it is dropped (with an error message if `gnrc_icmpv6_error` is compiled in). "Forwarding to an application" is meant in the way of resolving the `(nettype, demux_ctx)` tuple. So @Lotterleben was right, but not only with regards to "firewalling". I will however replace `application` with `upper layer` ;-).

File: sys/include/net/gnrc/ipv6.h (hunk 7 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: The application makes itself suitable by registering itself. A person developing for GNRC or reading this sentence should know what this mean.

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: But here as well: I will replace application with "upper layer". I guess I was in an NDN mindset, when I was writing this :sweat_smile:

File: sys/include/net/gnrc/ipv6.h (hunk 10 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Again. A reader should know this when following the reference above.

File: sys/include/net/gnrc/ipv6.h (hunk 12 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Why should the addresses be rewritten for a loopback (which what this section describes)?

File: sys/include/net/gnrc/ipv6.h (hunk 13 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Tried to but I think this makes it more complicated than it needs to. Essentially: default values, user config or in case of the source address derived from interface configuration and destination address.

File: sys/include/net/gnrc/ipv6.h (hunk 5 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Ah, but as a remark it fits. Sorry, Github was showing this weirdly.

File: sys/include/net/gnrc/ipv6.h (hunk 10 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Will add it as a remark however.

**Comment by miri64** [comment]
Addressed comments

File: sys/include/net/gnrc/ipv6.h (hunk 4 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Done

File: sys/include/net/gnrc/ipv6.h (hunk 7 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Did that with the rework of the sentence requested in https://github.com/RIOT-OS/RIOT/pull/10583#discussion_r243770438

File: sys/include/net/gnrc/ipv6.h (hunk 8 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Broke it up to be easier to understand

File: sys/include/net/gnrc/ipv6.h (hunk 11 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Changed to *must* ;-)

File: sys/include/net/gnrc/ipv6.h (hunk 12 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Added the words "loopback address" and "looped back" to make this clearer

File: sys/include/net/gnrc/ipv6.h (hunk 14 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Done

File: sys/include/net/gnrc/ipv6.h (hunk 16 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Done

File: sys/include/net/gnrc/ipv6.h (hunk 17 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Done

File: sys/include/net/gnrc/ipv6.h (hunk 18 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Done

File: sys/include/net/gnrc/ipv6.h (hunk 19 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Added links to the `gnrc_pktbuf` functions these words resolve to ;-)

File: sys/include/net/gnrc/ipv6.h (hunk 12 above)
Comment (tcschmidt) [reply sys/include/net/gnrc/ipv6.h]: I didn't see loopback here, but understood source routing on the same node. If source routing is meant , we need a re-addressing. If the destination address is (and exclusively was) the loopback address, then we are simply traversing (partially down) the stack and up again.

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Why exclusively loopback? If the destination address is assigned to an interface of the sending node the same applies.

File: sys/include/net/gnrc/ipv6.h (hunk 15 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Oops, overlooked this one.

File: sys/include/net/gnrc/ipv6.h (hunk 17 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Done

File: sys/include/net/gnrc/ipv6.h (hunk 15 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: Fixed now.

**Comment by Lotterleben** [comment]
@miri64 yes please (if none of the other reviewers have any remaining issues?)

**Comment by miri64** [comment]
Squashed

**Comment by Lotterleben** [comment]
If I understand Travis right, this PR seems to introduce a Doxygen warning though, could you fix that?

**Comment by miri64** [comment]
Yepp, will fix, when I have Internet on my laptop again :)

**Comment by miri64** [comment]
All checks passed.

**Comment by miri64** [comment]
@smlng can we merge?

File: sys/include/net/gnrc/ipv6.h (hunk 2 above)
Comment (smlng) [reply sys/include/net/gnrc/ipv6.h]: bc it's in the IPv6 header file and not in `sys/include/net/gnrc/netapi.h`. If this is actually the NETAPI documentation it should go into that file and not here ... at least that seems to be sensible to me?

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: But that's exactly why I think it doesn't need more specification. This is the documentation of netapi commands that IPv6 accepts (not netapi documentation). The fact that this is IPv6-related is clear through the fact that it is in this file / group. Or does it really seem unclear for you in [the 6LoWPAN equivalent](http://doc.riot-os.org/group__net__gnrc__sixlowpan.html) (which is up for years already).

Comment (Lotterleben) [reply sys/include/net/gnrc/ipv6.h]: How about rewording the heading to something less ambiguous? Maybe...
```suggestion
 * # Supported NETAPI commands
```

Comment (smlng) [reply sys/include/net/gnrc/ipv6.h]: @miri64 ah okay, I think I get your point now ... I wasn't aware of the 6lo doc (at least that it is rather similar). However, what I saw was that the NETAPI module does not have such (extensive) documentation, hence I thought it might be misplaced here and should go in that header. 

TL;DR: leave as here, but might be nice to add some generic documentation/text on NETAPI in the respective header (later), too.

Comment (smlng) [reply sys/include/net/gnrc/ipv6.h]: On another note: @Lotterleben makes sense to me anyway.

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: > TL;DR: leave as here, but might be nice to add some generic documentation/text on NETAPI in the respective header (later), too.

[The generic GNRC doc](http://doc.riot-os.org/group__net__gnrc.html) contains a lot about NETAPI. Since NETAPI is a pretty central part to GNRC, I don't know how much sense it makes to redundantly add extensive documentation to NETAPI then.

> On another note: @Lotterleben's comment makes sense to me anyway.

Since that also requires a change to 6LoWPAN to keep them in sync, I rather would vote to do that in a separate PR. How about that?

Comment (smlng) [reply sys/include/net/gnrc/ipv6.h]: sounds good to me.

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: > Since that also requires a change to 6LoWPAN to keep them in sync, I rather would vote to do that in a separate PR. How about that?

See https://github.com/RIOT-OS/RIOT/pull/10718
